spatial/r3/transform.go

```diff
@@ -0,0 +1,267 @@
+		t.x03*t.x12*t.x20*t.x31 + t.x03*t.x12*t.x21*t.x30
+}
+
+// Inv returns the inverse of the transform such that
+// t.Inv() * t is the identity Transform.
+// If matrix is singular then Inv() returns the zero transform.
+func (t Transform) Inv() Transform {
+	if t == (Transform{}) {
+		return t
+	}
+	det := t.Det()
+	if math.Abs(det) < 1e-16 {
+		return zeroTransform
+	}
+	// Do something if singular?
+	d := 1 / det
+	x00 := t.d00 + 1
+	x11 := t.d11 + 1
+	x22 := t.d22 + 1
+	x33 := t.d33 + 1
+	var m Transform
+	m.d00 = (t.x12*t.x23*t.x31-t.x13*x22*t.x31+t.x13*t.x21*t.x32-x11*t.x23*t.x32-t.x12*t.x21*x33+x11*x22*x33)*d - 1
+	m.x01 = (t.x03*x22*t.x31 - t.x02*t.x23*t.x31 - t.x03*t.x21*t.x32 + t.x01*t.x23*t.x32 + t.x02*t.x21*x33 - t.x01*x22*x33) * d
+	m.x02 = (t.x02*t.x13*t.x31 - t.x03*t.x12*t.x31 + t.x03*x11*t.x32 - t.x01*t.x13*t.x32 - t.x02*x11*x33 + t.x01*t.x12*x33) * d
+	m.x03 = (t.x03*t.x12*t.x21 - t.x02*t.x13*t.x21 - t.x03*x11*x22 + t.x01*t.x13*x22 + t.x02*x11*t.x23 - t.x01*t.x12*t.x23) * d
+	m.x10 = (t.x13*x22*t.x30 - t.x12*t.x23*t.x30 - t.x13*t.x20*t.x32 + t.x10*t.x23*t.x32 + t.x12*t.x20*x33 - t.x10*x22*x33) * d
+	m.d11 = (t.x02*t.x23*t.x30-t.x03*x22*t.x30+t.x03*t.x20*t.x32-x00*t.x23*t.x32-t.x02*t.x20*x33+x00*x22*x33)*d - 1
+	m.x12 = (t.x03*t.x12*t.x30 - t.x02*t.x13*t.x30 - t.x03*t.x10*t.x32 + x00*t.x13*t.x32 + t.x02*t.x10*x33 - x00*t.x12*x33) * d
+	m.x13 = (t.x02*t.x13*t.x20 - t.x03*t.x12*t.x20 + t.x03*t.x10*x22 - x00*t.x13*x22 - t.x02*t.x10*t.x23 + x00*t.x12*t.x23) * d
+	m.x20 = (x11*t.x23*t.x30 - t.x13*t.x21*t.x30 + t.x13*t.x20*t.x31 - t.x10*t.x23*t.x31 - x11*t.x20*x33 + t.x10*t.x21*x33) * d
+	m.x21 = (t.x03*t.x21*t.x30 - t.x01*t.x23*t.x30 - t.x03*t.x20*t.x31 + x00*t.x23*t.x31 + t.x01*t.x20*x33 - x00*t.x21*x33) * d
+	m.d22 = (t.x01*t.x13*t.x30-t.x03*x11*t.x30+t.x03*t.x10*t.x31-x00*t.x13*t.x31-t.x01*t.x10*x33+x00*x11*x33)*d - 1
+	m.x23 = (t.x03*x11*t.x20 - t.x01*t.x13*t.x20 - t.x03*t.x10*t.x21 + x00*t.x13*t.x21 + t.x01*t.x10*t.x23 - x00*x11*t.x23) * d
+	m.x30 = (t.x12*t.x21*t.x30 - x11*x22*t.x30 - t.x12*t.x20*t.x31 + t.x10*x22*t.x31 + x11*t.x20*t.x32 - t.x10*t.x21*t.x32) * d
+	m.x31 = (t.x01*x22*t.x30 - t.x02*t.x21*t.x30 + t.x02*t.x20*t.x31 - x00*x22*t.x31 - t.x01*t.x20*t.x32 + x00*t.x21*t.x32) * d
+	m.x32 = (t.x02*x11*t.x30 - t.x01*t.x12*t.x30 - t.x02*t.x10*t.x31 + x00*t.x12*t.x31 + t.x01*t.x10*t.x32 - x00*x11*t.x32) * d
+	m.d33 = (t.x01*t.x12*t.x20-t.x02*x11*t.x20+t.x02*t.x10*t.x21-x00*t.x12*t.x21-t.x01*t.x10*x22+x00*x11*x22)*d - 1
+	return m
+}
+
+// transpose returns the transposed Transform.
+func (t Transform) transpose() Transform {
+	return Transform{
+		d00: t.d00, x01: t.x10, x02: t.x20, x03: t.x30,
+		x10: t.x01, d11: t.d11, x12: t.x21, x13: t.x31,
+		x20: t.x02, x21: t.x12, d22: t.d22, x23: t.x32,
+		x30: t.x03, x31: t.x13, x32: t.x23, d33: t.d33,
+	}
+}
+
+// equals tests the equality of the Transforms to within a tolerance.
+func (t Transform) equals(b Transform, tolerance float64) bool {
```

Comment:
I think this belongs in the test file. I'd also refer to use a more standard distance measure. If `Transform` implemented `mat.Matrix` this would just fall out.